docs/theories/datatypes.rst

```diff
@@ -0,0 +1,229 @@
+Theory Reference: Datatypes
+===========================
+
+Logic
+-----
+
+To enable cvc5's decision procedure for datatypes, include ``DT`` in the logic:
+
+.. code:: smtlib
+
+  (set-logic QF_UFDT)
+
+Alternatively, use the ``ALL`` logic:
+
+.. code:: smtlib
+
+  (set-logic ALL)
+
+Syntax
+------
+
+cvc5 supports the following syntax for declaring mutually recursive blocks of
+datatypes in ``*.smt2`` input files in the smt lib 2.6 format:
+
+.. code:: smtlib
+
+  (declare-datatypes ((D1 n1) ... (Dk nk))
+   (((C1 (S1 T1) ... (Si Ti)) ... (Cj ... ))
+    ...
+    ((...) ... (...)))
+
+where ``D1 ... Dk`` are datatype types, ``C1 ... Cj`` are the constructors for
+datatype ``D1``,
+``S1 ... Si`` are the selectors (or "destructors") of constructor ``C1``, and
+each ``T1 ... Ti`` is a previously declared type or one of ``D1 ... Dk``.
+The symbols ``U1 ... Un`` are type parameters (fresh symbols).
+The numbers ``n1 ... nk`` denote the number of type
+parameters for the datatype, where ``0`` is used for non-parametric datatypes.
+
+In addition to declaring symbols for constructors and selectors, the above
+command also adds tester (or "discriminator") indexed symbols of the form
+``(_ is C)`` for each constructor ``C``, which are unary predicates which
+evaluate to true iff their argument has top-symbol ``C``.
+
+Semantics
+---------
+
+The decision procedure for inductive datatypes can be found
+`here <http://homepage.cs.uiowa.edu/~tinelli/papers/BarST-JSAT-07.pdf>`__.
+
+Example Declarations
+--------------------
+
+An enumeration:
+
+.. code:: smtlib
+
+  (declare-datatypes ((Color 0))
+   (((Red) (Black))))
+
+A List of Int with ``cons`` and ``nil`` as constructors:::
+
+.. code:: smtlib
+
+  (declare-datatypes ((list 0))
+   (((cons (head Int) (tail list)) (nil))))
+
+A parametric List of T's:::
```

Comment:
The triple colons seem to mess with the Sphinx markup...
```suggestion
A List of Int with ``cons`` and ``nil`` as constructors:

.. code:: smtlib

  (declare-datatypes ((list 0))
   (((cons (head Int) (tail list)) (nil))))

A parametric List of T's:
```

Reply:
Yeah, that was unintended. Thanks for catching.